pkg/frontend/reader/merge/merge.go

```diff
@@ -0,0 +1,623 @@
+		}
+
+		return func() error {
+			affected, err := fl.merger.AddTraces(thumbnails)
+			if err != nil {
+				return err
+			}
+
+			for key := range affected {
+				fl.scheduleFrom(fl.merger.objects[key], followLimit, linkSelector)
+			}
+
+			return nil
+		}, nil
+	})
+}
+
+type ListFunc[M any] func(
+	ctx context.Context,
+	key objKey,
+	startTime, endTime time.Time,
+	limit int,
+) ([]TraceWithMetadata[M], error)
+
+func (merger *Merger[M]) FollowLinks(
+	ctx context.Context,
+	linkSelector tfconfig.LinkSelector,
+	startTime, endTime time.Time,
+	lister ListFunc[M],
+	concurrency int,
+	limit int32,
+	limitIsGlobal bool,
+) error {
+	fl := &followLinkPool[M]{
+		sem:       semaphore.New(concurrency),
+		knownKeys: sets.New[objKey](),
+		lister:    lister,
+		startTime: startTime,
+		endTime:   endTime,
+		merger:    merger,
+	}
+
+	for _, obj := range merger.objects {
+		fl.knownKeys.Insert(obj.key)
+	}
+
+	globalLimit := new(atomic.Int32)
+	globalLimit.Store(limit)
+
+	for _, obj := range merger.objects {
+		var remainingLimit *atomic.Int32
+		if limitIsGlobal {
+			remainingLimit = globalLimit
+		} else {
+			remainingLimit = new(atomic.Int32)
+			remainingLimit.Store(limit)
+		}
+
+		fl.scheduleFrom(obj, remainingLimit, linkSelector)
+	}
+
+	if err := fl.sem.Run(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (merger *Merger[M]) MergeTraces() ([]*MergeTree[M], error) {
+	abLinks := abLinkMap{}
+
+	for _, obj := range merger.objects {
+		for _, link := range obj.links {
+			abLink := abLinkFromTargetLink(obj.key, link)
+			abLinks.insert(abLink)
+		}
+	}
+
+	connectedComps := merger.findConnectedComponents(merger.objects, abLinks)
+	mergeTrees := make([]*MergeTree[M], 0, len(connectedComps))
+	for _, keys := range connectedComps {
+		var members []*object[M]
+		for _, key := range keys {
+			members = append(members, merger.objects[key])
+		}
+
+		mergeTree, err := newMergeTree(members, abLinks)
+		if err != nil {
+			return nil, err
+		}
+
+		mergeTrees = append(mergeTrees, mergeTree)
+	}
+
+	return mergeTrees, nil
+}
+
+type object[M any] struct {
+	key      objKey
+	metadata []M
+	tree     *tftree.SpanTree
+
+	links []TargetLink
+}
+
+func newObject[M any](key objKey, trace *tftree.SpanTree, metadata M) (*object[M], error) {
+	clonedTree, err := trace.Clone()
+	if err != nil {
+		return nil, fmt.Errorf("clone spans: %w", err)
+	}
+	obj := &object[M]{
+		key:      key,
+		metadata: []M{metadata},
+		tree:     clonedTree,
+	}
+	return obj, nil
+}
+
+func (obj *object[M]) merge(trace *tftree.SpanTree, metadata M) error {
+	obj.metadata = append(obj.metadata, metadata)
+
+	mergeRoot(obj.tree.Root, trace.Root)
+
+	copyVisitor := &copyTreeVisitor{to: obj.tree, toParent: obj.tree.Root.SpanID}
+	trace.Visit(copyVisitor)
+	if copyVisitor.err != nil {
+		return copyVisitor.err
+	}
+
+	return nil
+}
+
+func mergeRoot(base *model.Span, tail *model.Span) {
+	mergeRootInterval(base, tail)
+	mergeRootTags(base, tail)
```

Comment:
Supposedly the root span is the object pseudospan that does not have any logs right now. But sure, let's add it in case we add root span logs in the future.